address_to_census_tract.py

```diff
@@ -7,81 +7,167 @@
             "latlng": latlng,
-            "tract": tract.get("properties", {}).get("GEOID") if tract else None,
+            "tract": tract,
         }
 
         print(json.dumps({ **record, "result": result }))
 
-
-def address_to_latlng(address):
-    """Convert an address string to a list of latitude, longitude coordinates.
-
-    Currently uses Google's geocoder API for geocoding, but this could be
-    replaced with other geocoder implementations.
+def process_csv(file_path: str):
     """
-    return geocoder.google(address).latlng
-
+    Given a *file_path* to a CSV, processes the `address` column and adds 
+    extra columns for a standardized address, latitude and longitude
+    coordinates, and census tract. If a given address is invalid, these new 
+    columns are left blank. Saves the new table at `data/test/out.csv`. 
 
-def load_geojson(geojson_filename):
-    """Read GeoJSON file and return a list of features converted to shapes."""
+    To minimize costs, an address should only be looked up once (via 
+    `lookup_address()`).
+    """
+    df = pd.read_csv(file_path)
+    tracts = load_geojson("data/geojsons/Washington_2016.geojson")
 
-    with open(geojson_filename) as file:
-        geojson = json.load(file)
+    response = pd.Series(df['address'].apply(lookup_address))
 
-    return [
-        { "properties": feature["properties"], "shape": shape(feature['geometry']) }
-            for feature in geojson["features"]
-    ]
+    # Parse response object to create columns of interest
+    df['standard_address'] = response.apply(lambda x: x['standard_address'])
+    df['lat'] = response.apply(lambda x: x['lat'])
+    df['lng'] = response.apply(lambda x: x['lng'])
 
+    latlng = pd.Series(list(zip(df['lat'], df['lng'])))
+    df['final_tract'] = latlng.apply(lambda x: latlng_to_polygon(x, tracts))
+    
+    df.to_csv('data/test/out.csv') 
 
-def latlng_to_polygon(latlng, polygons):
+def lookup_address(address: str) -> dict:
     """
-    Find the first polygon in *polygons* which contains the *latlng* and return
-    the polygon, else None.
+    Given an address, returns a dict containing a standardized address and 
+    lat/long coordinates from SmartyStreet's US Street geocoding API.
     """
+    auth_id = os.environ['SMARTYSTREETS_AUTH_ID']
+    auth_token = os.environ['SMARTYSTREETS_AUTH_TOKEN']
+
+    credentials = StaticCredentials(auth_id, auth_token)
+    client = ClientBuilder(credentials).build_us_street_api_client()
+
+    lookup = Lookup()
+    lookup.street = address
+    lookup.candidates = 1
+    lookup.match = "Invalid"  # Most permissive
+    
+    try:
+        client.send_lookup(lookup)
+    except exceptions.SmartyException as err:
+        print(err)
+        return
+
+    result = lookup.result
+    if not result:  # Invalid address
+        result = extract_address(credentials, address)
+
+    first_candidate = result[0]
+    return {"standard_address": first_candidate.delivery_line_1 + ' ' + 
```

Comment:
What about `delivery_line_2`?

Reply:
I wasn't sure if we needed that level of granularity, but I can easily add it back in.